### PR TITLE
Java: Improve config discovery and misconfiguration reporting

### LINKED ISSUE #2093: Java: Display a clearer message when properties files is not found
To get the settings from a properties file, you use the following code

```java
AzureOpenAISettings settings = AIProviderSettings.getAzureOpenAISettingsFromFile("conf.properties");
```

If something goes wrong the following message is displayed:

```
java.io.IOException: conf.properties not configured properly
```

Because the code swallows the exception, I had to debug the code of `ClientSettings` to see that the property file was not found (`java.io.FileNotFoundException: /agoncal/agoncal-sample-semantic-kernel/conf.properties (No such file or directory)`):

```java
    protected String getSettingsValueFromFile(
            String settingsFile, String property, String clientSettingsId) throws IOException {
        File Settings = new File(settingsFile);

        try (FileInputStream fis = new FileInputStream(Settings.getAbsolutePath())) {
            Properties props = new Properties();
            props.load(fis);

            return props.getProperty(getPropertyNameForClientId(clientSettingsId, property));
        } catch (IOException e) {
            throw new IOException(settingsFile + " not configured properly");
        }
    }
```

 It would be clearer, either if you display the exception message, or if you handle `FileNotFoundException` :

```java

       // Display exception message
        } catch (IOException e) {
            throw new IOException(settingsFile + " not configured properly:" + e.getMessage());
        }

       // Handle FileNotFoundException
        } catch (FileNotFoundException e) {
            throw new FileNotFoundException(settingsFile + " not found");
        } catch (IOException e) {
            throw new IOException(settingsFile + " not configured properly");
        }
```

### PR DESCRIPTION
**Pull request by johnoliver**
Move much of the logic of OpenAI config loading from samples to connectors

Fixes #2093

### CODE REVIEW

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientProvider.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+                    return buildOpenAiClient();
+                case AZURE_OPEN_AI:
+                    return buildAzureOpenAiClient();
+                default:
+                    throw new ConfigurationException(NoValidConfigurationsFound);
+            }
+        }
+    }
+
+    private OpenAIAsyncClient buildOpenAiClient() throws ConfigurationException {
```
Comment (brunoborges) [inline java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientProvider.java:233]: Method name should be `buildOpenAIClient`.

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientProvider.java [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+            LOGGER.warn(e.getMessage());
+            throw e;
+        }
+
+        return new OpenAIClientBuilder()
+                .credential(new NonAzureOpenAIKeyCredential(settings.getKey()))
+                .buildAsyncClient();
+    }
+
+    private OpenAIAsyncClient buildAzureOpenAiClient() throws ConfigurationException {
```
Comment (brunoborges) [inline java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientProvider.java:248]: Method name should be `buildAzureOpenAIClient`.

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/SettingsMap.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+ * </ul>
+ */
+public class SettingsMap {
+    private static final Logger LOGGER = org.slf4j.LoggerFactory.getLogger(SettingsMap.class);
+
+    public static final String CONF_PROPERTIES_NAME = "CONF_PROPERTIES";
+    public static final String CONF_PROPERTIES =
+            getEnvOrProperty(CONF_PROPERTIES_NAME, "conf.properties");
+
+    private static final List<File> DEFAULT_PROPERTIES_LOCATIONS =
```
Comment (brunoborges) [inline java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/SettingsMap.java:46]: Should the instantiation of this be lazy loaded?

**Review by brunoborges — APPROVED** [review]
Overall this looks great! Besides the inline comments, I've got one extra question:

We may want to give a mechanism for developers to load the configuration from the classpath, in case they choose to bundle `conf.properties` inside their application JAR file. I recall another PR or change that did allow this.

But I guess we can wait for user feedback on this extra configuration option.

**Review by brunoborges — CHANGES_REQUESTED** [review]
Couple other things:

I think the class `OpenAIClientSettings` should be called `AbstractOpenAIClientSettings` to avoid confusion with `OpenAISettings`.

And the documentation of the order of loading settings should also be at the class type level javadoc of AbstractClientSettings, but perhaps in a simpler way, pointing the user to read SettingsMap documentation, so we don't have too much documentation duplicated everywhere.

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientSettings.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    /**
+     * Check if the settings are valid
+     *
+     * @return true if the settings are valid
+     */
+    public abstract boolean assertIsValid() throws ConfigurationException;
+
+    /**
+     * Get the Azure OpenAI key
```
Comment (brunoborges) [inline java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientSettings.java:18]: Doc should say "Get the client key"

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientSettings.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) Microsoft. All rights reserved.
+package com.microsoft.semantickernel.connectors.ai.openai.util;
+
+import com.microsoft.semantickernel.exceptions.ConfigurationException;
+
+public abstract class OpenAIClientSettings {
```
Comment (brunoborges) [inline java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientSettings.java:6]: May want to add a javadoc stating that this class is primarily for `OpenAI` and `Azure OpenAI` clients.

**Review by brunoborges — CHANGES_REQUESTED** [review]

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/AzureOpenAISettings.java [hunk 6]
```diff
@@ -67,55 +55,26 @@ public String getDeploymentName() {
...
-                        path, Property.AZURE_OPEN_AI_DEPLOYMENT_NAME.label(), clientSettingsId);
-        return this;
-    }
-
-    @Override
-    public boolean isValid() {
-        return key != null && endpoint != null;
+    public String getKey() {
+        if (key == null) {
+            throw new RuntimeException("Azure OpenAI key is not set");
```
[the remark this answers is not in the document]
Comment (johnoliver) [reply java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/AzureOpenAISettings.java:60]: I have moved it to IllegalState, making it a checked exception would make it quite awkward, and really by the time this function is used, it should already be validated

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/exceptions/ConfigurationException.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft. All rights reserved.
+package com.microsoft.semantickernel.exceptions;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/** Kernel logic exception */
+public class ConfigurationException extends Exception {
```
Comment (dsgrieve) [inline java/semantickernel-api/src/main/java/com/microsoft/semantickernel/exceptions/ConfigurationException.java:8]: Should extend com.microsoft.semantickernel.SKException

File: java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/SettingsMap.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+                                LOGGER.info("Failed to read config file at: {}", file);
+                            }
+                        } else {
+                            LOGGER.info("Did not find configuration file: {}", file);
+                        }
+                    });
+        }
+
+        // Overlay system properties
+        properties.putAll(System.getProperties());
```
[the remark this answers is not in the document]
Comment (brunoborges) [reply java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/SettingsMap.java:141]: After we filter out the more Spring/Servlet/JNDI specific ways of configuring an app (while keeping some that are interesting for our case), this is what Spring devs are used to:

1. Config data (such as application.properties files).
2. OS environment variables.
3. Java System properties (System.getProperties()).
4. Properties from SPRING_APPLICATION_JSON (inline JSON embedded in an environment variable or system property).
5. Command line arguments.
6. Application properties outside of your packaged jar (application.properties including YAML and profile variants).
7. Application properties packaged inside your jar (application.properties including YAML and profile variants).
8. Default properties (specified using SpringApplication.setDefaultProperties).
9. [Devtools global settings properties](https://docs.spring.io/spring-boot/docs/current/reference/html/using.html#using.devtools.globalsettings) in the $HOME/.config/spring-boot directory when devtools is active.

[Spring Boot 3.x current](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config)

## Proposal

Translating the above to Java SK, the values would to be considered in the following order: 

_Note: later options overrides the values of earlier ones)_

1. Config data (`META-INF/sk/config.properties`).
1. OS environment variables
1. Java System properties (e.g., `-Dclient.key...=value`)
1. Inline JSON object (either in `SK_CONFIG_JSON` environment variable or `-Dsk.config.json=<value>`)
1. Developer environment location (`$HOME/.sk/config.properties`)

As for the properties, I propose the following names (for both properties and system environment variables, as well as a JSON object):

```text
sk.client.openai.key=SK_CLIENT_OPENAI_KEY

sk.client.azureopenai.key=SK_CLIENT_AZUREOPENAI_KEY
sk.client.azureopenai.endpoint=SK_CLIENT_AZUREOPENAI_ENDPOINT
sk.client.azureopenai.deploymentname=SK_CLIENT_AZUREOPENAI_DEPLOYMENTNAME
```

Thoughts?

@johnoliver @dsgrieve

File: java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/AzureOpenAISettings.java [hunk 9]
```diff
@@ -62,60 +50,31 @@ public String getEndpoint() {
      */
     public String getDeploymentName() {
         if (deploymentName == null) {
-            throw new RuntimeException("Azure OpenAI deployment name is not set");
+            throw new IllegalStateException("Azure OpenAI deployment name is not set");
```
[the remark this answers is not in the document]
Comment (brunoborges) [reply java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/AzureOpenAISettings.java:53]: I don't think it is.

File: java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/AzureOpenAISettings.java [hunk 10]
```diff
@@ -50,7 +38,7 @@ public String getKey() {
      */
     public String getEndpoint() {
         if (endpoint == null) {
-            throw new RuntimeException("Azure OpenAI endpoint is not set");
+            throw new IllegalStateException("Azure OpenAI endpoint is not set");
```
[the remark this answers is not in the document]
Comment (johnoliver) [reply java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/AzureOpenAISettings.java:41]: Making it a checked exception would be quite messy for calling code at the moment, and generally the config _should_ have been asserted valid before this is called (at least it is in our code)

File: java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientProvider.java [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+     * instantiate an OpenAI client, and if that fails, then it will attempt to instantiate an Azure
+     * OpenAI client.
+     *
+     * @return A client instance
+     * @throws ConfigurationException If the settings are not found
+     */
+    public OpenAIAsyncClient getAsyncClient() throws ConfigurationException {
+        // No client type is specified, search for a valid client type
+        if (clientType == null) {
+            LOGGER.info("No OpenAI client type specified, searching for a valid client type");
```
[the remark this answers is not in the document]
Comment (johnoliver) [reply java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientProvider.java:185]: There is quite a lot of logging, we have had a few issues with people not quite understanding where the config is getting loaded from, so I want to be fairly clear where things are getting loaded from, certainly can drop this down to debug

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/exceptions/ConfigurationException.java (hunk 7 above)
Comment (johnoliver) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/exceptions/ConfigurationException.java:8]: I would prefer this to be a checked exception...maybe we should have a CheckedSKException

File: java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/AzureOpenAISettings.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (dsgrieve) [reply java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/AzureOpenAISettings.java:41]: NoSuchElementException is a RuntimeException (as is IllegalStateException).

That the user should validate the settings before calling these methods is bound to cause problems. An option here would be to use a static method instead which could do the validate and throw the ConfigurationException.

File: java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientProvider.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (dsgrieve) [reply java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/OpenAIClientProvider.java:185]: Logging an info when all alternatives fail is appropriate.

File: java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/SettingsMap.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (johnoliver) [reply java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/SettingsMap.java:141]: - At least for now, not supporting inline JSON as we dont have JSON config at all so I would remove 4
- Seems odd to me that Spring would have `$HOME/.sk/config.properties` as the highest priority, IMO it should be the lowest, reading the docs I think they only use it for configuring the dev tools and not the application. Either way, imo move 5 down to 1

[the remark this answers is not in the document]
Comment (brunoborges) [reply java/connectors/semantickernel-connectors-ai-openai/src/main/java/com/microsoft/semantickernel/connectors/ai/openai/util/SettingsMap.java:141]: Updated the proposal with this feedback. Thanks @johnoliver

**Review by dsgrieve — APPROVED** [review]

**Pull request merged**